### PR TITLE
[Gemma3] compile ✨ 

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Enables compilation on Gemma3 (and re-enables it on Gemma2 / Cohere2).

Reverts #36620 
Supercedes #37433 (solves the same problem, but this PR is much cleaner)

### Performance
Measured on an RTX4090, excluding compile warmup time:
- [Example script from gemma 3 4b](https://huggingface.co/google/gemma-3-4b-it#running-the-model-on-a-singlemulti-gpu):  3.87s on `main` (not compiled) -> 2.39s this PR
- [Example script from gemma 2 9b](https://huggingface.co/google/gemma-2-9b-it#running-the-model-on-a-single--multi-gpu): 3.67s on `main` (not compiled) -> 2.18s this PR

### Tests
- [x] slow gemma 2 tests (9 failing tests from `main` -> need to be revisited)
- [x] slow gemma 3 tests (2 failing tests from `main`, `tests/models/gemma3/test_modeling_gemma3.py::Gemma3Vision2TextModelTest::test_eager_matches_sdpa_generate` gets fixed in this PR)

### Post-mortem: How did we break compile on Gemma 2?
1. Doing `git bisect`, compilation first "breaks" in the PR where the cache is initialized in the `meta` device (https://github.com/huggingface/transformers/pull/35164). "break" here doesn't mean "crash", but rather "becomes very slow". Curiously, this change doesn't slow down `StaticCache` + `llama` (why?), so it flew under the radar when we benchmarked before merging. Nevertheless, this specific PR has been reverted (https://github.com/huggingface/transformers/pull/36543).
2. Along the way, we corrected how the sliding window attention works, by slicing the attention mask correctly (https://github.com/huggingface/transformers/pull/35681). However, the solution here is not `torch.compile` friendly: `forward` now has an `int` argument that is different at each forward pass at generation time, causing recompilation ([reference](https://pytorch.org/docs/stable/torch.compiler_troubleshooting.html#wrapping-constants-with-tensors)). The changes in this PR work around this issue.

### CODE REVIEW

File: src/transformers/models/cohere2/modeling_cohere2.py [hunk 1]
```diff
@@ -426,73 +430,6 @@ def _init_weights(self, module):
                 module.weight.data[module.padding_idx].zero_()
 
 
-COHERE2_INPUTS_DOCSTRING = r"""
```
Comment (gante) [inline src/transformers/models/cohere2/modeling_cohere2.py:429]: the diff in the docstring is not intentional.

I'm assuming that `modular` has seen logic changes since this file was last auto-generated (we skip models wo diff). If anyone knows how remove this from this PR's diff, lmk 🙏

File: src/transformers/models/cohere2/modeling_cohere2.py [hunk 2]
```diff
@@ -528,8 +465,7 @@ def get_input_embeddings(self):
     def set_input_embeddings(self, value):
         self.embed_tokens = value
 
-    @can_return_tuple
-    @add_start_docstrings_to_model_forward(COHERE2_INPUTS_DOCSTRING)
```
Comment (gante) [inline src/transformers/models/cohere2/modeling_cohere2.py:532]: same here 🤔 

if any of you knows the right solution lmk, otherwise I'll have a look :D

File: src/transformers/models/cohere2/modular_cohere2.py [hunk 3]
```diff
@@ -400,11 +396,16 @@ def forward(
...
                 # Should only be used when beyond the sliding window (i.e. offset > 0)
                 offset = max(0, offset)
-                attention_mask = attention_mask[:, :, :, offset : offset + effective_seq_len]
+                # equivalent to: `attention_mask = attention_mask[:, :, :, offset : offset + effective_seq_len]`,
+                # but without data-dependent slicing (i.e. torch.compile friendly)
+                mask_indexes = torch.arange(
+                    min(effective_seq_len, attention_mask.shape[-1]), device=attention_mask.device
+                )
+                mask_indexes += offset
+                attention_mask = attention_mask[:, :, :, mask_indexes]
```
Comment (gante) [inline src/transformers/models/cohere2/modular_cohere2.py:408]: Core change for the PR.

`attention_mask = attention_mask[:, :, :, offset : offset + effective_seq_len]` requires either passing an integer in the signature to build `offset` (previous solution, not compile friendly) or doing data-dependent slicing (crashes compile)

The solution is to:
1. build a fixed-shape `arange` -- we can use `shapes` to create compile-compatible arrays on the fly ✅ 
2. add some tensor to a tensor ✅ 
3. slice a tensor with another tensor ✅ 

(Note: at first I tried `torch.roll` + fixed-shape slicing, but `torch.roll` doesn't support the argument `shifts=offset`, `shifts` has to be an integer 😢 )

File: src/transformers/models/cohere2/modeling_cohere2.py [hunk 4]
```diff
@@ -345,11 +344,16 @@ def forward(
...
+                offset = cache_position[-1] - effective_seq_len
                 # Should only be used when beyond the sliding window (i.e. offset > 0)
                 offset = max(0, offset)
-                attention_mask = attention_mask[:, :, :, offset : offset + effective_seq_len]
+                # equivalent to: `attention_mask = attention_mask[:, :, :, offset : offset + effective_seq_len]`,
+                # but without data-dependent slicing (i.e. torch.compile friendly)
+                mask_indexes = torch.arange(
+                    min(effective_seq_len, attention_mask.shape[-1]), device=attention_mask.device
+                )
+                mask_indexes += offset
```
Comment (ArthurZucker) [inline src/transformers/models/cohere2/modeling_cohere2.py:355]: are you sure this is cuda graph compatible?~

File: src/transformers/cache_utils.py [hunk 5]
```diff
@@ -1652,9 +1652,7 @@ class HybridCache(Cache):
         ```
     """
 
-    # TODO (joao): dive deeper into gemma2 and paligemma -- there are reports of speed loss with compilation. Revert
-    # ALL changes from the PR that commented the line below when reactivating it.
-    # is_compileable = True
+    is_compileable = True
```
Comment (ArthurZucker) [inline src/transformers/cache_utils.py:1655]: nice! Can we update the cache to also init the layers lazily like we dofor HybridChunked cache?

File: src/transformers/models/cohere2/modeling_cohere2.py (hunk 4 above)
Comment (gante) [reply src/transformers/models/cohere2/modeling_cohere2.py:355]: yes, see e.g. scripts at the top of the PR header

File: src/transformers/cache_utils.py (hunk 5 above)
Comment (gante) [reply src/transformers/cache_utils.py:1655]: @ArthurZucker `HybridChunkedCache` only works if we don't compile the first forward pass, `HybricCache` works regardless of we compile the first forward pass or not. `torch._dynamo.mark_static_address` can't be called inside `torch.compile`, which lazy init does.

This means that if a user creates their own custom code with `HybridChunkedCache`, they can't simply compile the forward pass. If anything, `HybridChunkedCache` should move away from lazy init :P

File: src/transformers/models/cohere2/modular_cohere2.py [hunk 6]
```diff
@@ -351,6 +348,7 @@ def __init__(self, config: Cohere2Config, layer_idx: int):
         self.is_sliding = (layer_idx + 1) % self.config.sliding_window_pattern != 0
         self.sliding_window = config.sliding_window
 
+    @deprecate_kwarg("last_cache_position", version="4.53.0")
```
Comment (qubvel) [inline src/transformers/models/cohere2/modular_cohere2.py:351]: Modular files inherit decorators for overwritten methods unless you add any new decorator on top of it; that's why you have decorators disappeared in modeling file

Comment (qubvel) [reply src/transformers/models/cohere2/modular_cohere2.py:351]: The solution is to manually add all disappeared decorators 🥲

File: src/transformers/cache_utils.py (hunk 5 above)
Comment (gante) [reply src/transformers/cache_utils.py:1655]: Chatted offline: 
1. lazy init is needed for TP
2. however, lazy init is incompatible with compiling the first forward pass (prefill). lazy init + `@torch.compiler.disable() `doesn't solve it either
3. solution: add a new flag `lazy_init = None`. If `torch.distributed` is initialized and the flag is unset, then it will be `True`.
4. Apply this change to ALL caches -> ALL caches compatible with TP

**Review by zucchini-nlp — APPROVED** [review]
Nice, thanks for detailed investigation! Do you think we can add a slow test to compare generation time with compile or include `HybridCache` in benchmarks board, so we don't accidentally introduce graph breaks? Given that Gemma3 is a high usage model and supports only Hybrid cache, I think it's important to not break it

**Review by Cyrilvallez — COMMENTED** [review]
Hey! Super nice work, thanks a lot! Just the offset that is wrong by 1, I added a detailed comment about that.

Also, as I assume you're already aware, this is only working under the assumption of a single prefill step with more than 1 tokens, then only decoding steps with 1 tokens. This was bothering me for some time I must say (what about context caching???), but never took the time to change it to make it more general and fullproof. However, with llama4, as I added prefill chunking, I had to make it work for any situation and any number of new tokens at any time. This is a nice precedent that I think we should use to now make the sliding caches more general. 
However, this should probably be a separate PR, if you plan on working on it at some point let me know!! 🤗

**Comment by gante** [comment]
> Do you think we can add a slow test to compare generation time with compile or include HybridCache in benchmarks board, so we don't accidentally introduce graph breaks? Given that Gemma3 is a high usage model and supports only Hybrid cache, I think it's important to not break it

@zucchini-nlp Benchmarks are hard to put as a test: different devices/versions -> different speeds, needs multiple runs to avoid being flaky 💔  However, we should test that compilation only happens once for the entire forward pass, i.e. that there are no graph breaks nor recompilations. I'm going to explore torch docs to see if we can add this to a test, in this PR. At the moment, we are often doubting the quality of our compiled forward passes, which is not great

**Comment by zucchini-nlp** [comment]
@gante yeah, we can do something similar to what diffusers is trying to do (in continuation to yesterday's thread discussions). Fine by me, as long as we check that there are no recompilations every step

**Comment by gante** [comment]
@Cyrilvallez off by 1 comment addressed 👍 LMK if you'd like any further change

@Cyrilvallez @zucchini-nlp There are three follow-up items to this PR, where each will be a separate PR:
1. Mixin test update to confirm that no recompilations, dynamic shapes, ... are happening. [Related comment](https://github.com/huggingface/transformers/pull/37447/#issuecomment-2812272952). [I've decided not to include it in this PR, since it needs to touch other models to fix failures];
2. TP + compile compatible caches. [Related comment](https://github.com/huggingface/transformers/pull/37447/#discussion_r2039726257)
3. Gemma 3 + Prefill chunking + more than one new token at generation time. [Related comment](https://github.com/huggingface/transformers/pull/37447/#discussion_r2044016869)

**Review by Cyrilvallez — APPROVED** [review]
LGTM!! 🤗 Thanks a lot, super glad to have compile back and to simplify by removing the extra arg that was introduced as well! 💛

**Comment by guangy10** [comment]
Hey @gante, can you point me how gemma3 is compiled, maybe a simple test? I looked through this PR but don't see a test for compile. BTW I'm working on add `torch.export` support for this model, and some changes in this PR are helpful with the export work as you are removing some of the data-dependent code which is also required by `torch.export`.

**Comment by guangy10** [comment]
FYI, in https://github.com/huggingface/transformers/pull/37728 `torch.export` is supported as well! AND the model can be lowered to ExecuTorch and run e2e w/o any issue!
